keps/sig-apps/0008-topology-aware-workload-controllers.md

```diff
@@ -0,0 +1,350 @@
+control all sorts of planned disruptions, including updates. 
+
+To achieve this, PDB should allow users to specify allowed pod disruptions for
+a collection of nodes (such as the nodes in one failure domain), using a new PDB
+field `.spec.nodeSelectorTerm`. 
+
+This requires changes to workloads controllers code (change pod deletions to
+evictions), PDB API, and PDB controller code.
+
+[`PodDisruptionBudget`]: https://kubernetes.io/docs/concepts/workloads/pods/disruptions/
+
+### User Stories
+
+#### Story 1: Update By Failure Domain 
+
+When a user wants to update one failure domain at a time, they can create a PDB
+to disallow any disruptions in failure domains other than the first domain to
+update:
+
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: app-pdb
+spec:
+  maxUnavailable: 0
+  selector:
+    matchLabels:
+      app: foo
+  nodeSelectorTerm:
+      # Only allow evicting/updating pods in zone “bar”
+      matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/zone
+        operator: NotIn
+        values: 
+        - bar
+      - key: failure-domain.beta.kubernetes.io/zone
+        operator: Exists
+```
+
+#### Story 2: Update Across Failure Domain (PDB Per Failure Domain)
+
+When a user wants to update across failure domains, and only allow one pod per
+failure domain to be unavailable at a time, they can specify:
+
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: app-pdb
+spec:
+  maxUnavailable: 1
+  selector:
+    matchLabels:
+      app: foo
+  nodeSelectorTerm:
+      # Only allow 1 pod to be evicted/updated per zone
+      matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/zone
+        operator: Exists
+```
+
+#### Story 3: Update Across Failure Domain (Overall PDB)
+
+When a user wants to update across failure domains and only allow one pod to be
+unavailable at a time in all failure domains, they can specify:
+
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: app-pdb
+spec:
+  maxUnavailable: 1
+  selector:
+    matchLabels:
+      app: foo
+```
+
+#### Story 4: Scale Down 
+
+When a user wants to scale down a workload controller, they can also use PDB to
+limit overall disruptions or in specific failure domains, similar to Story 3. 
+
+### Implementation Details/Notes/Constraints
+
+* We don't support more than 1 PDB rule covering the same pod today. This needs
+  to be updated. 
+* The topology-aware PDB's maxUnavailable and minAvaiable is calculated based on
+  (number of replicas / number of domains).
+* Update one failure domain at a time is manual. It might be fine because it’s
+  an advanced use case.
+
+### Risks and Mitigations
+
+Controllers will be changed to evict pods instead of delete pods.
+
+Risks:
+
+* If a pod eviction fails because of PDB, the controller may enter eviction
+  hotloop.
+* The controller may create more API calls than it should, because the
```

Comment:
If I am getting it correctly, this makes me worried about performance, maybe the approach itself if the workload controller will be fighting with PDB with etcd in between. (Succeeding eventually, but producing unnecessary load.)